common-pg/src/main/scala/com/socrata/pg/analyzer2/rollup/RollupExact.scala

```diff
@@ -0,0 +1,814 @@
+package com.socrata.pg.analyzer2.rollup
+
+import org.slf4j.LoggerFactory
+
+import com.socrata.soql.analyzer2._
+import com.socrata.soql.collection.OrderedMap
+
+import com.socrata.pg.analyzer2.{SqlizerUniverse, Stringifier}
+
+object RollupExact {
+  private val log = LoggerFactory.getLogger(classOf[RollupExact[_]])
+}
+
+class RollupExact[MT <: MetaTypes](
+  semigroupRewriter: SemigroupRewriter[MT],
+  functionSubset: FunctionSubset[MT],
+  functionSplitter: FunctionSplitter[MT],
+  splitAnd: SplitAnd[MT],
+  stringifier: Stringifier[MT]
+) extends ((Select[MT], RollupInfo[MT], LabelProvider) => Option[Statement[MT]]) with SqlizerUniverse[MT] {
+  import RollupExact.log
+
+  private type IsoState = IsomorphismState.View[MT]
+
+  // see if there's a rollup that can be used to answer _this_ select
+  // (not any sub-parts of the select!).  This needs to produce a
+  // statement with the same output schema (in terms of column labels
+  // and types) as the given select.
+  def apply(select: Select, rollupInfo: RollupInfo[MT], labelProvider: LabelProvider): Option[Statement] = {
+    if(select.hint(SelectHint.NoRollup)) {
+      log.debug("Bailing because query asked us not to roll it up")
+      return None
+    }
+
+    val candidate = rollupInfo.statement match {
+      case sel: Select =>
+        sel
+      case stmt =>
+        log.debug("Bailing because rollup is not a Select:\n  {}", stringifier.statement(stmt).indent(2))
+        return None
+    }
+
+    log.debug(
+      "Attempting to rewrite:\n  QUERY:\n    {}\n  ROLLUP:\n    {}",
+      stringifier.statement(select).indent(4): Any,
+      stringifier.statement(rollupInfo.statement).indent(4)
+    )
+
+    val isoState = select.from.isomorphicTo(candidate.from).getOrElse {
+      log.debug("Bailing because the rollup's From is not the same as the query's From")
+      return None
+    }
```

Comment:
This is technically stronger than it needs to be, but weakening it to be more inclusive is subtle and since it's not super invasive that can be done Later.

Reply:
I'd prefer this comment to be in the code, Just In Case™, but it's not a big deal.